bin/ch-completion.bash

```diff
@@ -0,0 +1,388 @@
+            compopt -o nospace
+        fi
+        ;;
+    push )
+        if [[ "$prev" == "--image" ]]; then
+            compopt -o nospace
+            COMPREPLY=( $(compgen -d -S / -- $cur) )
+            return 0
+        # The following check seems to fix a bug where the completion
+        # function initialzes an empty storage directory.
+        elif [[ -n "$(ls "$strg_dir"/img)" ]]; then
+            COMPREPLY=( $(compgen -W "$($CH_BIN/ch-image list -s $strg_dir) --image" -- "$cur") )
+            _ltrim_colon_completions "$cur"
+        fi
+        ;;
+    undelete )
+        # FIXME: Update with “ch-image undelete --list” once #1551 drops
+        COMPREPLY=()
+        ;;
+    '' )
+        # Only autocomplete subcommands if there's no subcommand present.
+        COMPREPLY=( $(compgen -W "$_image_subcommands" -- $cur) )
+        ;;
+    esac
+
+    # If we've made it this far, the last remaining option for completion
+    # is common opts. Note that we do the “-n” check to avoid being overzealous
+    # with our suggestions.
+    if [[ -n "$cur" ]]; then
+        COMPREPLY+=( $(compgen -W "$_image_common_opts" -- $cur) )
+        return 0
+    fi
+}
+
+## ch-run ##
+
+# Completion function for ch-run
+#
+_ch_run_completion () {
+    echo "does nothing"
+}
+
+## Helper functions ##
+
+# Use this function to enable or disable completion after sourcing
+# this file. Useful if there are issues.
+# FIXME: Add other executables as their completion functions are
+#        implemented.
+ch-completion () {
+    if [[ "$1" == "disable" ]]; then
+        complete -r ch-image
+    elif [[ "$1" == "enable" ]]; then
+        complete -F _ch_image_completion ch-image
+    fi
+}
+
+# Figure out which storage directory to use (including cli-specified
+# storage).
+# FIXME: Can probably cook up a sed pattern that'll remove the need
+#        for the “if” statement.
+_ch_find_storage () {
+    if [[ -n "$(grep -Eo "(\-\-storage|[^\-]\-s)" <<< "$@")" ]]; then
+        # This sed only works as desired if “--storage” or “-s” are present
+        # in the command line...
+        sed -E 's/(.*)(--storage|[^-]-s)\ *([^ ]*)(.*$)/\3/g' <<< "$@"
+    elif [[ -n "$CH_IMAGE_STORAGE" ]]; then
+        echo "$CH_IMAGE_STORAGE"
+    else
+        echo "/var/tmp/$USER.ch/"
+    fi
+}
+
+# Horrible, disgusting function to find an image or image ref in
+# the ch-run command line.
+#
+# NOT FINISHED, DON'T USE!!!
+_ch_run_image_finder () {
+    # Takes array of words. Tries to find an image in there.
+    #local -n words=$1
+    local wrds=("$@")
+    local ct=1
+
+    #echo "len: ${#wrds[@]}"
+
+    while (($ct < ${#wrds[@]})); do
+        if [[ ( -f "${wrds[$ct]}" && ( "${wrds[$ct]}" == *".sqfs" || "${wrds[$ct]}" == *".tar."* || "${wrds[$ct]}" == *".tgz" ) ) || ( -d "${wrds[$ct]}" && "${wrds[$ct-1]}" != "--mount" &&
+                "${wrds[$ct-1]}" != "-m" && "${wrds[$ct-1]}" != "--bind" &&
+                "${wrds[$ct-1]}" != "-b" ) ]]; then
```

Comment:
This conditional is indeed quite rough. Maybe `case` could help? Or temporary variables instead of `${wrds[$ct]}` and `${wrds[$ct-1]}`?

In either case, I think you need a comment explaining what's going on here. It's something about files with certain extensions, or directories, unless the thing is preceded by `--bind` or `--mount`?

Or maybe you can leverage context better? There's only so many places in a `ch-run` command that there can be an image.

Indentation can help too (and you don't need quotes inside `[[ ]]`):
```suggestion
        if [[    (    -f ${wrds[$ct]} \
                   && (    ${wrds[$ct]} == *.sqfs \
                        || ${wrds[$ct]} == *.tar.? \
                        || ${wrds[$ct]} == *.tar.?? \
                        || ${wrds[$ct]} == *.tgz ) ) \
              || (    -d ${wrds[$ct]} \
                   && ${wrds[$ct-1]} != --mount \
                   && ${wrds[$ct-1]} != -m \
                   && ${wrds[$ct-1]} != --bind \
                   && ${wrds[$ct-1]} != -b ) ]]; then
```

There is not an agreed-on extension for SquashFS archives. So identifying those by extension is going to be very error-prone. Maybe `*.*sq*` (any extension containing `sq`)???